tuf_conformance/test_updater_key_rotations.py

```diff
@@ -0,0 +1,147 @@
+        MdVersion(keys=[5, 6, 7, 8, 9], threshold=3, sigs=[0, 2, 4, 5, 6, 7]),
+        MdVersion(keys=[5, 6, 7, 8, 9], threshold=3, sigs=[5, 6, 7]),
+    ],
+    "1-of-3-threshold-increase-to-2-of-3": [
+        MdVersion(keys=[1, 2, 3], threshold=1, sigs=[1]),
+        MdVersion(keys=[1, 2, 3], threshold=2, sigs=[1, 2]),
+    ],
+    "2-of-3-threshold-decrease-to-1-of-3": [
+        MdVersion(keys=[1, 2, 3], threshold=2, sigs=[1, 2]),
+        MdVersion(keys=[1, 2, 3], threshold=1, sigs=[1, 2]),
+        MdVersion(keys=[1, 2, 3], threshold=1, sigs=[1]),
+    ],
+    "1-of-2-threshold-increase-to-2-of-2": [
+        MdVersion(keys=[1], threshold=1, sigs=[1]),
+        MdVersion(keys=[1, 2], threshold=2, sigs=[1, 2]),
+    ],
+    "1-of-1-key-rotation-fail-not-signed-with-old-key": [
+        MdVersion(keys=[1], threshold=1, sigs=[1]),
+        MdVersion(keys=[2], threshold=1, sigs=[2, 3, 4], res=False),
+    ],
+    "1-of-1-key-rotation-fail-not-signed-with-new-key": [
+        MdVersion(keys=[1], threshold=1, sigs=[1]),
+        MdVersion(keys=[2], threshold=1, sigs=[1, 3, 4], res=False),
+    ],
+    "3-of-5-one-key-rotate-fails-not-signed-with-3-new-keys": [
+        MdVersion(keys=[0, 1, 2, 3, 4], threshold=3, sigs=[0, 2, 4]),
+        MdVersion(keys=[0, 1, 3, 4, 5], threshold=3, sigs=[0, 2, 4], res=False),
+    ],
+    "3-of-5-one-key-rotate-fails-not-signed-with-3-old-keys": [
+        MdVersion(keys=[0, 1, 2, 3, 4], threshold=3, sigs=[0, 2, 4]),
+        MdVersion(keys=[0, 1, 3, 4, 5], threshold=3, sigs=[0, 4, 5], res=False),
+    ],
+    "1-of-3-threshold-bump-to-2-of-3-fails-new-threshold-not-reached": [
+        MdVersion(keys=[1, 2, 3], threshold=1, sigs=[1]),
+        MdVersion(keys=[1, 2, 3], threshold=2, sigs=[2], res=False),
+    ],
+    "2-of-3-threshold-decr-to-1-of-3-fails-old-threshold-not-eached": [
+        MdVersion(keys=[1, 2, 3], threshold=2, sigs=[1, 2]),
+        MdVersion(keys=[1, 2, 3], threshold=1, sigs=[1], res=False),
+    ],
+}
+
+ids = root_rotation_cases.keys()
+cases = root_rotation_cases.values()
+
+@pytest.mark.parametrize("root_versions", cases, ids=ids)
+def test_root_rotation(client: ClientRunner,
+                       server: SimulatorServer,
+                       root_versions: list[MdVersion]) -> None:
+    """Test client refresh with various sequences of root updates
+
+        Each MdVersion in root_versions describes root keys and signatures of a
+        remote root metadata version. As an example:
+            MdVersion([1,2,3], 2, [1,2])
+        defines a root that contains keys 1, 2 and 3 with threshold 2. The
+        metadata is signed with keys 1 and 2.
+
+        Assert that refresh result is expected and that local root on disk is
+        the expected one after all roots have been loaded from remote using the
+        standard client update workflow.
+        """
+
+    signers = []
+    for _ in range(10):
+        signer = CryptoSigner.generate_ed25519()
```

Comment:
can this be `generate_ecdsa()` for consistency with `RepositorySimulator#add_key`?

Reply:
I think something is dumping the ed25519 keys differently from the ecdsa ones:

```ruby
{:key_type=>"ecdsa", :schema=>"ecdsa-sha2-nistp256", :key_bytes=>"-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEu5zBhkn/oty+dBFz55Lw2MVBi4nl\nDS+lRRBX2RrJ+w4VFJT7QVjRwYHODVVjFl2eZyoH6GRN/4B7hO8CHa0L3g==\n-----END PUBLIC KEY-----\n", :key_id=>"06d227f3fa01231f2ca3d62e9e4d55b053f914dd7c849a21c46213ee72225d7d"}
```

vs

```ruby
{:key_type=>"ed25519", :schema=>"ed25519", :key_bytes=>"e02f66bb1667c902a5d96565807f609febf38b8335be7771c18d2c1806ffe8ed", :key_id=>"367d6b82ed9f08e94b74f49d075e698aca29bd693d94dc9b521f09bb67e229e8"}
```

Reply:
scratch that, that's what the spec (unexpectedly calls for). all figured out!

Reply:
> can this be `generate_ecdsa()` for consistency with `RepositorySimulator#add_key`?

yeah, it could, no need to be different just because. The different keytypes can be (or maybe are already?) tested in a separate test